feat(l10n): Allow to force language with a query parameter

come-nc opened the pull request:
* Resolves: # <!-- related github issue -->

## Summary

This will be used by support application to always generate the support
 report in english. It may also be used to debug a language, or to
 temporarily show a page in an other language to take a screenshot or to
 share your screen with someone speaking another language.

## TODO

- [ ] Should that be a header instead?
- [ ] Where should that be documented?
- [ ] Can it cause any troubles anywhere? I don’t think so :thinking: 

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

come-nc committed:
feat(l10n): Allow to force language with a query parameter

This will be used by support application to always generate the support
 report in english. It may also be used to debug a language, or to
 temporarily show a page in an other language to take a screenshot or to
 share your screen with someone speaking another language.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

nickvergessen commented on lib/private/L10N/Factory.php:
```diff
@@ -98,7 +98,7 @@ public function get($app, $lang = null, $locale = null) {
 			}
 
-			$forceLang = $this->config->getSystemValue('force_language', false);
+			$forceLang = $this->request->getParam('forceLanguage') ?? $this->config->getSystemValue('force_language', false);
```
Also should overwrite in those?
- getLanguageFromRequest
- findLanguage
- findGenericLanguage

come-nc committed:
feat(l10n): Respect forceLanguage query parameter in more places

Attempt to make L10N Factory consistent with itself when forceLanguage
 query parameter is used.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

nickvergessen commented on lib/private/L10N/Factory.php:
```diff
@@ -411,6 +411,10 @@ public function getUserLanguage(?IUser $user = null): string {
 				return $language;
 			}
 
+			if (($forcedLanguage = $this->request->getParam('forceLanguage')) !== null) {
```
In the other methods it takes priority over the config, should it also here?

come-nc replied:
I hesitated. force_language from configuration does takes precedence. But for forceLanguage query parameter it felt weird to take priority over user language when specifically asking for it.

But I’m not sure here, because if the calling code tries to get an L10N instance using that language it will still get the forced one, so I’m not sure whether it can be useful.

Basically the corner-case here is if I use ?forceLanguage=de and I share a file by email it may send the email in German regardless of the target language setting. But 1) either way is fine to me, it’s really a corner case and 2) it’s not really gonna do that any way because the request from frontend button clic will not have the query parameter.

So, tl;dr:
I’m fine with forcing language on first line of getUserLanguage as other methods do if you think it’s better.
If we do that, maybe we should even force it in getLanguages like the config setting is doing, making the config option and the query parameter consistent with each other.

nickvergessen replied:
lets do it and see where we end up.
Have to check that sending emails in target language doesn't fail, but I guess the amount of requests where we `?forceLanguage` will also be very limited.
